src/types.ts

```diff
@@ -0,0 +1,335 @@
+  doc: Document[] | string[],
+) => CallbackComponentReturnValue;
+
+export type IIndexInsertOrRemoveFunction<I> = (
+  index: I,
+  id: string,
+  prop: string,
+  value: SearchableValue,
+  language: string | undefined,
+  tokenizer: Tokenizer,
+  docsCount: number,
+) => CallbackComponentReturnValue;
+
+export type IIndexRemoveFunction<I> = (index: I, id: string, prop: string) => CallbackComponentReturnValue;
+
+export interface IIndex<S extends Schema, I extends OpaqueIndex, D extends OpaqueDocumentStore> {
+  create: (lyra: Lyra<S, I, D>, schema: Schema) => I;
+
+  beforeInsert?: IIndexInsertOrRemoveFunction<I>;
+  insert: IIndexInsertOrRemoveFunction<I>;
+  afterInsert?: IIndexInsertOrRemoveFunction<I>;
+
+  beforeRemove?: IIndexInsertOrRemoveFunction<I>;
+  remove: IIndexInsertOrRemoveFunction<I>;
+  afterRemove?: IIndexInsertOrRemoveFunction<I>;
+
+  search(index: I, prop: string, terms: string, context: SearchContext): CallbackComponentReturnValue<TokenScore[]>;
+  searchByWhereClause(index: I, filters: Record<string, boolean | ComparisonOperator>): string[];
+
+  getSearchableProperties(index: I): CallbackComponentReturnValue<string[]>;
+  getSearchablePropertiesWithTypes(index: I): CallbackComponentReturnValue<Record<string, SearchableType>>;
+
+  load(raw: unknown): I | Promise<I>;
+  save(index: I): unknown | Promise<object>;
+}
+
+export interface IDocumentsStore<S extends Schema, I extends OpaqueIndex, D extends OpaqueDocumentStore> {
+  create: (lyra: Lyra<S, I, D>) => D;
+  get(store: D, id: string): CallbackComponentReturnValue<Document | undefined>;
+  getMultiple(store: D, ids: string[]): CallbackComponentReturnValue<(Document | undefined)[]>;
+  store(store: D, id: string, doc: Document): CallbackComponentReturnValue<boolean>;
+  remove(store: D, id: string): CallbackComponentReturnValue<boolean>;
+  count(store: D): CallbackComponentReturnValue<number>;
+
+  load(raw: unknown): D | Promise<D>;
+  save(store: D): unknown | Promise<object>;
+}
+
+export interface Tokenizer {
+  tokenize: (raw: string, language?: string) => string[];
+}
+
+export interface ComplexComponent<S extends Schema, I extends OpaqueIndex, D extends OpaqueDocumentStore> {
+  tokenizer: Tokenizer;
+  index: IIndex<S, I, D>;
+  documentsStore: IDocumentsStore<S, I, D>;
+}
+
+export interface SimpleComponents {
+  validateSchema(doc: Document, schema: Schema): CallbackComponentReturnValue<boolean>;
+  getDocumentIndexId(doc: Document): CallbackComponentReturnValue<string>;
+  getDocumentProperties(
+    doc: Document,
+    paths: string[],
+  ): CallbackComponentReturnValue<Record<string, string | number | boolean>>;
+  formatElapsedTime(number: bigint): CallbackComponentReturnValue<bigint> | CallbackComponentReturnValue<string>;
+}
+
+export interface SimpleOrArrayCallbackComponents<
+  S extends Schema,
+  I extends OpaqueIndex,
+  D extends OpaqueDocumentStore,
+> {
+  beforeInsert: SingleCallbackComponent<S, I, D> | SingleCallbackComponent<S, I, D>[];
```

Comment:
Can we move this `SingleCallbackComponent<S, I, D> | SingleCallbackComponent<S, I, D>[];` into a type alias?

Reply:
Sure, I will.

Reply:
Done! Will be there in the next push.